firedrake/slate/slac/compiler.py

```diff
@@ -100,11 +113,67 @@ def compile_expression(slate_expr, tsfc_parameters=None):
     try:
         return cache[key]
     except KeyError:
-        kernel = SlateKernel(slate_expr, tsfc_parameters).split_kernel
+        kernel = SlateKernel(slate_expr, tsfc_parameters, coffee).split_kernel
         return cache.setdefault(key, kernel)
 
 
+def generate_loopy_kernel(slate_expr, tsfc_parameters=None):
+    cpu_time = time.time()
+    # TODO: Get PyOP2 to write into mixed dats
+    if slate_expr.is_mixed:
+        raise NotImplementedError("Compiling mixed slate expressions")
+
+    if len(slate_expr.ufl_domains()) > 1:
+        raise NotImplementedError("Multiple domains not implemented.")
+
+    Citations().register("Gibson2018")
+
+    print("COMPILING SLATE", slate_expr)
+    # Create a loopy builder for the Slate expression,
+    # e.g. contains the loopy kernels coming from TSFC
+    builder = LocalLoopyKernelBuilder(expression=slate_expr,
+                                      tsfc_parameters=tsfc_parameters)
+
+    print("BUILDER DONE")
+
+    # Stage 1: slate to gem....
+    gem_expr = slate_to_gem(builder)
+
+    # Stage 2a: gem to loopy...
+    loopy_outer = gem_to_loopy(gem_expr, builder)
+
+    # Stage 2b: merge loopys...
+    loopy_merged = merge_loopy(loopy_outer, builder.templated_subkernels, builder)  # builder owns the callinstruction
+    print("LOOPY KERNEL GLUED")
+
+    # Stage 2c: register callables...
+    loopy_merged = get_inv_callable(loopy_merged)
+    loopy_merged = get_solve_callable(loopy_merged)
+
+    # WORKAROUND: Generate code directly from the loopy kernel here,
+    # then attach code as a c-string to the op2kernel
```

Comment:
This came from the sloopy branch here https://github.com/firedrakeproject/firedrake/blob/a80c98eaa26b13981357304784378de24fc56289/firedrake/slate/slac/compiler.py#L573.
I think when we started talking about this, you mentioned that loopy cannot do "stacked inlining". I guess if you are in PyOP2 and do tensor assembly via TSFC that is fine, because you inline only once, but now we inline the TSFC kernel in the Slate kernel and that would again need to be inlined in PyOP2. But maybe loopy fixed that by now, I can check later.

Reply:
I have had a try at avoiding this which I addressed in https://github.com/OP2/PyOP2/pull/580.